scipy/optimize/least_squares.py

```diff
@@ -0,0 +1,657 @@
+
+    if jac not in ['2-point', '3-point'] and not callable(jac):
+        raise ValueError("`jac` must be '2-point', '3-point' or callable.")
+
+    scaling = check_scaling(scaling, x0)
+
+    ftol, xtol, gtol = check_tolerance(ftol, xtol, gtol)
+
+    if not in_bounds(x0, lb, ub):
+        raise ValueError("`x0` is infeasible.")
+
+    def fun_wrapped(x):
+        return np.atleast_1d(fun(x, *args, **kwargs))
+
+    f0 = fun_wrapped(x0)
+    if f0.ndim != 1:
+        raise RuntimeError("`fun` must return at most 1-d array_like.")
+
+    n = x0.size
+    m = f0.size
+    if method == 'lm' and m < n:
+        raise ValueError("Method 'lm' doesn't work when the number of "
+                         "residuals is less than the number of variables.")
+
+    if jac in ['2-point', '3-point']:
+        if jac_sparsity is not None:
+            if method == 'lm':
+                warn("`jac_sparsity` is ignored for method='lm', dense "
+                     "differencing will be used.")
+                sparsity = None
+            else:
+                sparsity = check_jac_sparsity(jac_sparsity, n, m)
+        else:
+            sparsity = None
+
+        def jac_wrapped(x, f):
+            J = approx_derivative(
+                fun, x, rel_step=diff_step, method=jac, f0=f, bounds=bounds,
+                args=args, kwargs=kwargs, sparsity=sparsity)
+
+            if issparse(J) and tr_solver == 'exact':
+                warn("Jacobian is converted to dense for tr_solver='exact' "
+                     "even though `jac_sparsity` is provided. Consider using "
+                     "'lsmr' trust-region solver or set `jac_sparsity` to "
+                     "None.")
+                J = J.toarray()
+            elif not issparse(J):
+                J = np.atleast_2d(J)
+
+            return J
+    else:
+        def jac_wrapped(x, _=None):
+            J = jac(x, *args, **kwargs)
+
+            if issparse(J):
+                if method == 'lm':
+                    warn("Jacobian is converted to dense for "
+                         "method='lm', consider using a different `method` "
+                         "or return dense Jacobian.")
+                    J = J.toarray()
+                elif tr_solver == 'exact':
+                    warn("Jacobian is converted to dense for "
+                         "tr_solver='exact', consider using 'lsmr' "
+                         "trust-region solver or return dense Jacobian.")
+                    J = J.toarray()
+                else:
+                    J = csr_matrix(J)
+            elif not isinstance(J, LinearOperator):
+                J = np.atleast_2d(J)
+
+            return J
+
+    J0 = jac_wrapped(x0, f0)
+    if len(J0.shape) != 2:
+        raise RuntimeError("`jac` must return at most 2-d array_like, "
+                           "sparse matrix or LinearOperator.")
+
+    if J0.shape != (m, n):
+        raise RuntimeError("The return value of `jac` has wrong shape.")
```

Comment:
Can also combine both checks. From a user perspective, I'd prefer to see an exact shape in the error message.